core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala

```diff
@@ -615,7 +615,11 @@ private[spark] class ExecutorAllocationManager(
       val taskIndex = taskEnd.taskInfo.index
       val stageId = taskEnd.stageId
       allocationManager.synchronized {
-        numRunningTasks -= 1
+        if (numRunningTasks > 0) {
```

Comment:
Hm, this seems like a band-aid though. Is this because the task-end event happens in the wrong order with the job end event? can we catch and guard for that case more directly?